src/courses/views.py

```diff
@@ -207,22 +212,65 @@ def get_success_url(self):
+class CourseStudentCreate(NonPublicOnlyViewMixin, SuccessMessageMixin, LoginRequiredMixin, UserPassesTestMixin, CreateView):
+
+    # if an active CourseStudent object assigned to student exists when student accesses registration view (user already enrolled), page will 403
+    def test_func(self):
+        return not CourseStudent.objects.filter(user=self.request.user, active=True).exists()
+
     model = CourseStudent
     form_class = CourseStudentForm
     # fields = ['semester', 'block', 'course', 'grade']
     success_url = reverse_lazy('quests:quests')
     success_message = "You have been added to the %(course)s course"
 
+    def get(self, request, *args, **kwargs):
+        # when accessing this view, check if we need a form at all, or can just create the studentcourse object using all defaults
+        # if simplified_course_registration is enabled in siteconfig and all three form fields are hidden, object should be created automatically
+        simpleregistration = SiteConfig.get().simplified_course_registration
+        form = self.get_form()
+        form_auto_create_condition = (
+            simpleregistration and
+            form.fields['semester'].widget.input_type == 'hidden' and
+            form.fields['block'].widget.input_type == 'hidden' and
+            form.fields['course'].widget.input_type == 'hidden'
+        )
+
+        if form_auto_create_condition:
+            # form_auto_create_condition == True means only 1 active Semester, Block, Course object exists
+            # can get active objects directly with first() and siteconfig.active_semester
+            obj, created = CourseStudent.objects.get_or_create(
+                    user=self.request.user,
+                    semester=SiteConfig.get().active_semester,
+                    block=Block.objects.filter(active=True).first(),
+                    course=Course.objects.filter(active=True).first()
+            )
+
+            # after object has been created, redirect to normal success url and display normal success message
+            # fstring used instead of self.success_message because form context not available
+            if created:
+                messages.success(request, f"You have been added to the {obj.course} Course")
+            return redirect(self.success_url)
+        else:
+            return super().get(request, *args, **kwargs)
+
     def get_form_kwargs(self):
         kwargs = super(CreateView, self).get_form_kwargs()
-        kwargs['instance'] = CourseStudent(user=self.request.user)
+        # visible fields with defaults set via kwargs have their defaults overwritten whenever a form is manually submitted, no conditional needed
+        kwargs['student_registration'] = True
+        kwargs['instance'] = CourseStudent(user=self.request.user,
+                                           semester=SiteConfig.get().active_semester,
+                                           block=Block.objects.filter(active=True).first(),
```

Comment:
I think this is why the form isn't defaulting to the blank option, because you are setting the defaults here with .first().

Do you have to pass more than just the user?  That's how it used to work. The rest of the logic should be taken care of in the form itself.

Reply:
Those kwarg defaults are necessary to set values for hidden form fields, but on their own they do interfere with the field defaults when not hidden. Conditional added for block and course kwarg defaults so they'll appear as normal when not necessary.
![image](https://github.com/bytedeck/bytedeck/assets/105619909/0e908e05-3b45-4621-839b-86fd595d23e0)
note form shows blank option for field with multiple choices while still hiding fields as needed

Reply:
FYI, I appreciate the explanation. Makes review so much easier!